File: Assets/Scripts/Game/MergeSystem.cs
using System;
using System.Collections.Generic;
using UnityEngine;
using HexaMerge.Core;

namespace HexaMerge.Game
{
    public struct MergeResult
    {
        public bool Success;
        public HexCoord MergeTargetCoord;
        public int ResultValue;
        public int MergedCount;
        public int ScoreGained;
        public List<HexCoord> MergedCoords;
        public List<int> StepValues;
        public List<List<HexCoord>> DepthGroups;
    }

    public class MergeSystem
    {
        private readonly HexGrid grid;

        public MergeSystem(HexGrid grid)
        {
            this.grid = grid;
        }

        public List<HexCell> FindConnectedGroup(HexCoord startCoord)
        {
            var result = new List<HexCell>();
            HexCell startCell = grid.GetCell(startCoord);
            if (startCell == null || startCell.IsEmpty) return result;

            int targetValue = startCell.TileValue;
            var visited = new HashSet<HexCoord>();
            var queue = new Queue<HexCoord>();

            queue.Enqueue(startCoord);
            visited.Add(startCoord);

            while (queue.Count > 0)
            {
                HexCoord current = queue.Dequeue();
                HexCell cell = grid.GetCell(current);
                result.Add(cell);

                var neighbors = current.GetNeighbors();
                for (int i = 0; i < neighbors.Count; i++)
                {
                    HexCoord nCoord = neighbors[i];
                    if (visited.Contains(nCoord)) continue;

                    HexCell nCell = grid.GetCell(nCoord);
                    if (nCell == null || nCell.TileValue != targetValue) continue;

                    visited.Add(nCoord);
                    queue.Enqueue(nCoord);
                }
            }

            return result;
        }

        public MergeResult TryMerge(HexCoord tapCoord)
        {
            var result = new MergeResult
            {
                Success = false,
                MergeTargetCoord = tapCoord,
                MergedCoords = new List<HexCoord>()
            };

            HexCell tapCell = grid.GetCell(tapCoord);
            if (tapCell == null || tapCell.IsEmpty) return result;

            int baseValue = tapCell.TileValue;

            // BFS with depth tracking (트리 구조 형태로 거리 계산)
            var visited = new HashSet<HexCoord>();
            var queue = new Queue<HexCoord>();
            var depthMap = new Dictionary<HexCoord, int>();

            queue.Enqueue(tapCoord);
            visited.Add(tapCoord);
            depthMap[tapCoord] = 0;

            while (queue.Count > 0)
            {
                HexCoord current = queue.Dequeue();
                int currentDepth = depthMap[current];

                var neighbors = current.GetNeighbors();
                for (int i = 0; i < neighbors.Count; i++)
                {
                    HexCoord nCoord = neighbors[i];
                    if (visited.Contains(nCoord)) continue;

                    HexCell nCell = grid.GetCell(nCoord);
                    if (nCell == null || nCell.TileValue != baseValue) continue;

                    visited.Add(nCoord);
                    depthMap[nCoord] = currentDepth + 1;
                    queue.Enqueue(nCoord);
                }
            }

            int totalCells = visited.Count;
            if (totalCells < 2) return result;

            // 깊이별 그룹 생성 (depth 0 = tapCoord 제외)
            var depthGroupsMap = new Dictionary<int, List<HexCoord>>();
            int maxDepth = 0;

            foreach (var kvp in depthMap)
            {
                if (kvp.Value == 0) continue; // skip tap cell
                if (!depthGroupsMap.ContainsKey(kvp.Value))
                    depthGroupsMap[kvp.Value] = new List<HexCoord>();
                depthGroupsMap[kvp.Value].Add(kvp.Key);
                if (kvp.Value > maxDepth) maxDepth = kvp.Value;
            }

            // 깊이 레벨 수 = 값 계산 기준
            // 다른 깊이에 있는 블럭만 계산에 포함
            // 동일한 깊이에 있는 블럭은 계산에서 제외 (하나의 단계로 처리)
            int depthLevels = depthGroupsMap.Count;
            int mergedValue = CalculateMergeValueByDepth(baseValue, depthLevels);

            // StepValues (깊이 레벨당 1개, 단계별 2배 증가)
            var stepValues = new List<int>();
            int stepValue = baseValue;
            for (int d = 0; d < depthLevels; d++)
            {
                stepValue = Mathf.Min(stepValue * 2, TileHelper.MaxValue);
                stepValues.Add(stepValue);
            }

            // 모든 셀 클리어 후 타겟에 최종값 설정
            foreach (var coord in visited)
            {
                grid.GetCell(coord).Clear();
            }
            tapCell.SetValue(mergedValue);

            // DepthGroups 리스트 (깊이 깊은 것부터 = 가장 먼 블럭부터)
            var depthGroups = new List<List<HexCoord>>();
            for (int d = maxDepth; d >= 1; d--)
            {
                if (depthGroupsMap.ContainsKey(d))
                    depthGroups.Add(depthGroupsMap[d]);
            }

            // MergedCoords (tapCoord 첫 번째, 나머지 깊이 깊은 순)
            var mergedCoords = new List<HexCoord>();
            mergedCoords.Add(tapCoord);
            for (int i = 0; i < depthGroups.Count; i++)
            {
                mergedCoords.AddRange(depthGroups[i]);
            }

            result.Success = true;
            result.MergeTargetCoord = tapCoord;
            result.ResultValue = mergedValue;
            result.MergedCount = totalCells;
            result.ScoreGained = mergedValue * depthLevels;
            result.MergedCoords = mergedCoords;
            result.StepValues = stepValues;
            result.DepthGroups = depthGroups;

            return result;
        }

        public bool CanMerge(HexCoord coord)
        {
            HexCell cell = grid.GetCell(coord);
            if (cell == null || cell.IsEmpty) return false;
            return FindConnectedGroup(coord).Count >= 2;
        }

        private int CalculateMergeValueByDepth(int baseValue, int depthLevels)
        {
            long value = (long)baseValue;
            for (int i = 0; i < depthLevels; i++)
            {
                value *= 2;
                if (value >= TileHelper.MaxValue)
                    return TileHelper.MaxValue;
            }
            return (int)value;
        }
    }
}

File: src/core/MergeSystem.js
/**
 * @fileoverview Merge system handling BFS-based tile merging with depth tracking.
 * Ported from Unity C# MergeSystem.cs to pure JavaScript ES Module.
 */

import { HexCoord } from './HexCoord.js';
import { MAX_VALUE } from './TileHelper.js';

/**
 * @typedef {Object} MergeResult
 * @property {boolean} success - Whether the merge was successful
 * @property {HexCoord} tapCoord - The coordinate that was tapped
 * @property {number} baseValue - The original value of the tapped tile
 * @property {number} resultValue - The final merged value
 * @property {number} scoreGained - Points earned from this merge
 * @property {number} mergedCount - Total number of cells involved in the merge
 * @property {HexCoord[]} mergedCoords - All merged coordinates (tap coord first, then deepest-first)
 * @property {number[]} stepValues - Intermediate values at each depth step
 * @property {HexCoord[][]} depthGroups - Groups of coords by depth (deepest first)
 * @property {Map<string, string>} parentMap - Child coord key -> parent coord key (BFS tree)
 */

/**
 * Handles tile merge operations on a HexGrid using BFS with depth tracking.
 */
export class MergeSystem {
    /**
     * @type {import('./HexGrid.js').HexGrid}
     * @private
     */
    _grid;

    /**
     * @param {import('./HexGrid.js').HexGrid} grid
     */
    constructor(grid) {
        this._grid = grid;
    }

    /**
     * Attempt a merge starting from the tapped coordinate.
     * Uses BFS to find all connected cells with the same value,
     * tracks depth for value calculation, and applies the merge.
     *
     * @param {HexCoord} tapCoord
     * @returns {MergeResult|null} null if merge is not possible
     */
    tryMerge(tapCoord) {
        const tapCell = this._grid.getCell(tapCoord);
        if (!tapCell || tapCell.isEmpty) return null;

        const baseValue = tapCell.value;

        // BFS with depth tracking + parent tracking
        const visited = new Set();
        const queue = [];
        /** @type {Map<string, number>} coord key -> depth */
        const depthMap = new Map();
        /** @type {Map<string, string>} child key -> parent key */
        const parentMap = new Map();

        const startKey = tapCoord.toKey();
        queue.push(tapCoord);
        visited.add(startKey);
        depthMap.set(startKey, 0);

        while (queue.length > 0) {
            const current = queue.shift();
            const currentKey = current.toKey();
            const currentDepth = depthMap.get(currentKey);

            const neighbors = current.getNeighbors();
            for (let i = 0; i < neighbors.length; i++) {
                const nCoord = neighbors[i];
                const nKey = nCoord.toKey();

                if (visited.has(nKey)) continue;

                const nCell = this._grid.getCell(nCoord);
                if (!nCell || nCell.value !== baseValue) continue;

                visited.add(nKey);
                depthMap.set(nKey, currentDepth + 1);
                parentMap.set(nKey, currentKey);
                queue.push(nCoord);
            }
        }

        const totalCells = visited.size;
        if (totalCells < 2) return null;

        // Build depth groups (depth 0 = tapCoord, excluded from groups)
        /** @type {Map<number, HexCoord[]>} */
        const depthGroupsMap = new Map();
        let maxDepth = 0;

        for (const [key, depth] of depthMap) {
            if (depth === 0) continue; // skip tap cell
            if (!depthGroupsMap.has(depth)) {
                depthGroupsMap.set(depth, []);
            }
            depthGroupsMap.get(depth).push(HexCoord.fromKey(key));
            if (depth > maxDepth) maxDepth = depth;
        }

        // depthLevels = number of distinct BFS depth levels (used only for the
        // cascade animation grouping, not for the merge value).
        const depthLevels = depthGroupsMap.size;

        // Merge value follows the tree structure: each parent node that absorbs
        // its children multiplies the value — ×2 for a single child, ×4 when the
        // parent has 2 or more children. The total exponent is the sum of these
        // per-parent contributions (×2 = exponent 1, ×4 = exponent 2).
        const childCount = new Map();
        for (const [, parentKey] of parentMap) {
            childCount.set(parentKey, (childCount.get(parentKey) || 0) + 1);
        }
        let mergeExponent = 0;
        for (const [, count] of childCount) {
            mergeExponent += count >= 2 ? 2 : 1;
        }
        const mergedValue = this._calculateMergeValue(baseValue, mergeExponent);

        // StepValues: one per doubling step (drives intermediate count-up visuals)
        const stepValues = [];
        let stepValue = baseValue;
        for (let d = 0; d < mergeExponent; d++) {
            stepValue = Math.min(stepValue * 2, MAX_VALUE);
            stepValues.push(stepValue);
        }

        // Clear all merged cells, then set final value at tap position
        for (const key of visited) {
            const coord = HexCoord.fromKey(key);
            this._grid.getCell(coord).clear();
        }
        tapCell.setValue(mergedValue);

        // DepthGroups list: deepest first (reversed)
        const depthGroups = [];
        for (let d = maxDepth; d >= 1; d--) {
            if (depthGroupsMap.has(d)) {
                depthGroups.push(depthGroupsMap.get(d));
            }
        }

        // MergedCoords: tapCoord first, then remaining in deepest-first order
        const mergedCoords = [tapCoord];
        for (let i = 0; i < depthGroups.length; i++) {
            for (let j = 0; j < depthGroups[i].length; j++) {
                mergedCoords.push(depthGroups[i][j]);
            }
        }

        return {
            success: true,
            tapCoord,
            baseValue,
            resultValue: mergedValue,
            scoreGained: mergedValue * mergeExponent,
            mergedCount: totalCells,
            mergedCoords,
            stepValues,
            depthGroups,
            parentMap
        };
    }

    /**
     * Find all connected cells with the same value starting from a coordinate.
     * @param {HexCoord} startCoord
     * @returns {import('./HexCell.js').HexCell[]}
     */
    findConnectedGroup(startCoord) {
        const result = [];
        const startCell = this._grid.getCell(startCoord);
        if (!startCell || startCell.isEmpty) return result;

        const targetValue = startCell.value;
        const visited = new Set();
        const queue = [];

        const startKey = startCoord.toKey();
        queue.push(startCoord);
        visited.add(startKey);

        while (queue.length > 0) {
            const current = queue.shift();
            const cell = this._grid.getCell(current);
            result.push(cell);

            const neighbors = current.getNeighbors();
            for (let i = 0; i < neighbors.length; i++) {
                const nCoord = neighbors[i];
                const nKey = nCoord.toKey();

                if (visited.has(nKey)) continue;

                const nCell = this._grid.getCell(nCoord);
                if (!nCell || nCell.value !== targetValue) continue;

                visited.add(nKey);
                queue.push(nCoord);
            }
        }

        return result;
    }

    /**
     * Check whether a merge is possible at the given coordinate.
     * A merge requires at least 2 connected cells with the same value.
     * @param {HexCoord} coord
     * @returns {boolean}
     */
    canMerge(coord) {
        const cell = this._grid.getCell(coord);
        if (!cell || cell.isEmpty) return false;
        return this.findConnectedGroup(coord).length >= 2;
    }

    /**
     * Calculate the merged value: baseValue doubled `exponent` times.
     * exponent = floor(log2(totalMergedTiles)) per the benchmark rule.
     * @param {number} baseValue
     * @param {number} exponent - number of doublings
     * @returns {number}
     * @private
     */
    _calculateMergeValue(baseValue, exponent) {
        let value = baseValue;
        for (let i = 0; i < exponent; i++) {
            value *= 2;
            if (value >= MAX_VALUE) return MAX_VALUE;
        }
        return value;
    }
}
